bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/test/java/com/google/cloud/bigtable/mirroring/hbase1_x/asyncwrappers/TestMirroringResultScanner.java

```diff
@@ -0,0 +1,353 @@
+  @Test
+  public void testScannerCloseWhenBothCloseThrow() throws IOException {
+    ResultScanner primaryScannerMock = mock(ResultScanner.class);
+
+    VerificationContinuationFactory continuationFactoryMock =
+        mock(VerificationContinuationFactory.class);
+
+    AsyncResultScannerWrapper secondaryScannerWrapperMock = mock(AsyncResultScannerWrapper.class);
+    AsyncTableWrapper secondaryAsyncTableWrapperMock = mock(AsyncTableWrapper.class);
+    when(secondaryAsyncTableWrapperMock.getScanner(any(Scan.class)))
+        .thenReturn(secondaryScannerWrapperMock);
+
+    final ResultScanner mirroringScanner =
+        new MirroringResultScanner(
+            new Scan(),
+            primaryScannerMock,
+            secondaryAsyncTableWrapperMock,
+            continuationFactoryMock,
+            flowController,
+            new MirroringTracer(),
+            true);
+
+    doThrow(new RuntimeException("first")).when(primaryScannerMock).close();
+    doThrow(new RuntimeException("second")).when(secondaryScannerWrapperMock).asyncClose();
+
+    RuntimeException thrown =
+        assertThrows(
+            RuntimeException.class,
+            new ThrowingRunnable() {
+              @Override
+              public void run() {
+                mirroringScanner.close();
+              }
+            });
+
+    verify(primaryScannerMock, times(1)).close();
+    verify(secondaryScannerWrapperMock, times(1)).asyncClose();
+    assertThat(thrown).hasMessageThat().contains("first");
+    assertThat(thrown.getSuppressed()).hasLength(1);
+    assertThat(thrown.getSuppressed()[0]).hasMessageThat().contains("second");
+  }
+
+  @Test
+  public void testMultipleCloseCallsCloseScannersOnlyOnce() throws IOException {
+    ResultScanner primaryScannerMock = mock(ResultScanner.class);
+
+    VerificationContinuationFactory continuationFactoryMock =
+        mock(VerificationContinuationFactory.class);
+    AsyncResultScannerWrapper secondaryScannerWrapperMock = mock(AsyncResultScannerWrapper.class);
+    SettableFuture<Void> closedFuture = SettableFuture.create();
+    closedFuture.set(null);
+    when(secondaryScannerWrapperMock.asyncClose()).thenReturn(closedFuture);
+
+    AsyncTableWrapper secondaryAsyncTableWrapperMock = mock(AsyncTableWrapper.class);
+    when(secondaryAsyncTableWrapperMock.getScanner(any(Scan.class)))
+        .thenReturn(secondaryScannerWrapperMock);
+
+    final ResultScanner mirroringScanner =
+        new MirroringResultScanner(
+            new Scan(),
+            primaryScannerMock,
+            secondaryAsyncTableWrapperMock,
+            continuationFactoryMock,
+            flowController,
+            new MirroringTracer(),
+            true);
+
+    mirroringScanner.close();
+    mirroringScanner.close();
+    verify(primaryScannerMock, times(1)).close();
+    verify(secondaryScannerWrapperMock, times(1)).asyncClose();
+  }
+
+  @Test
+  public void testSecondaryNextsAreIssuedInTheSameOrderAsPrimary() throws IOException {
```

Comment:
Please add comments on this test, its not obvious to me why we need the ReverseOrderExectutorSerivce. This test can be more readable.

Reply:
Done.